feat(ssr): adds server cookie output via response

blakeoxx opened the pull request:
This brings the fix/feature completed in https://github.com/stevermeister/ngx-cookie-service/pull/315 to the v18 codebase and resolves https://github.com/stevermeister/ngx-cookie-service/issues/266 . It increases the `SsrCookieService`s ability to use the request and response objects provided by Express to read and write cookies on the server side. This also means that cookies written during SSR will be returned to the client in the response headers.

This PR also includes less refactors than the previous one, per maintainer request. The only refactors included in this PR were strictly necessary for enabling this feature.

abinvp commented:
When is this going to be merged? Seems like this is capable of solving some significant challenges.

pavankjadda commented:
@blakeoxx I tried this and didn't work as expected. For example, setting cookie in `SsrCookieServices` using the code does not set it. Do you have an example where it shows this code working?

bballardkele commented:
> @blakeoxx I tried this and didn't work as expected. For example, setting cookie in `SsrCookieServices` using the code does not set it. Do you have an example where it shows this code working?

I have a production repo this is working in, but I unfortunately can't share the proprietary code. Let me create a standalone test project for you.

blakeoxx commented:
@pavankjadda try out this test repo: https://github.com/blakeoxx/ngx-cookie-service-18-test. You can build and start the project with `npm run build` followed by `npm run serve:ssr:ssr-cookie-test`.

I generated this project with Angular's `ng new --ssr` command, added the scaffolding to server.ts as outlined in our README, and added a test to the AppComponent which outputs console logs on the server console and client/browser console to confirm the server-side cookie is readable and writable by both platforms. [Example output](https://imgur.com/a/TZzABpp).

Additionally, something I noticed was that NG18's SSR [defaults to using prerendering](https://angular.dev/guide/prerendering). I had to disable this [in the builder architect configuration](https://github.com/blakeoxx/ngx-cookie-service-18-test/commit/9316681ea2456974f0dbcb1439eaee896e5e34bd). It's intended behavior that SSR won't run on prerendered routes and not a bug of this code, but perhaps may be worth calling out in the README.

Let me know if you need anything else or don't see expected results.

betojsx commented:
This package seems to be highly needed within the Angular community, but I’ve noticed that this PR has been open for some time. I personally have a strong need for this functionality and tried to implement it in Angular 17, but it doesn’t seem to be working.

In my setup, I’m saving the token on the server using ngx-cookie-service, but when I try to retrieve it on the client side, the token is null.

@blakeoxx, do you happen to have any working demos for Angular 17 with a workflow similar to mine?

In advance, thanks for all your hard work on this project—it’s much appreciated!

betojsx commented:
I’d like to take this opportunity to share what I’m trying to achieve in my app, so other users with similar goals can benefit as well.

In my setup, I have a forwarded host from `server.ts`, and in my app, my goal is to:

1. Make a request to my backend based on the host received.
2. Save the token and some configuration data from the response to cookies using ngx-cookie-service-ssr.

I expect the browser to load this saved data without making new requests. However, when I try to retrieve the token and configuration, they’re all `null`.
 
 
 Here's my server.ts for reference

```ts
import 'zone.js/node';

import { APP_BASE_HREF } from '@angular/common';
import { CommonEngine } from '@angular/ssr';
import * as express from 'express';
import { existsSync } from 'node:fs';
import { join } from 'node:path';
import AppServerModule from './src/main.server';
import { REQUEST as SSR_REQUEST } from 'ngx-cookie-service-ssr';

// The Express app is exported so that it can be used by serverless Functions.
export function app(): express.Express {
    const server = express();
    const distFolder = join(process.cwd(), 'dist/tennistool/browser');
    const indexHtml = existsSync(join(distFolder, 'index.original.html'))
        ? join(distFolder, 'index.original.html')
        : join(distFolder, 'index.html');

    const commonEngine = new CommonEngine();

    server.set('view engine', 'html');
    server.set('views', distFolder);

    // Example Express Rest API endpoints
    // server.get('/api/**', (req, res) => { });
    // Serve static files from /browser
    server.get(
        '*.*',
        express.static(distFolder, {
            maxAge: '1y',
        })
    );

    // All regular routes use the Angular engine
    server.get('*', (req, res, next) => {
        const { protocol, originalUrl, baseUrl, headers } = req;

        commonEngine
            .render({
                bootstrap: AppServerModule,
                documentFilePath: indexHtml,
                url: `${protocol}://${headers.host}${originalUrl}`,
                publicPath: distFolder,
                providers: [
                    { provide: APP_BASE_HREF, useValue: baseUrl },
                    { provide: SSR_REQUEST, useValue: req },
                    { provide: 'RESPONSE', useValue: res },
                    
                    // harcoded host for testing
                    {
                        provide: 'X_FORWARDED_HOST',
                        useValue: 'sub.mydomain.com',
                    },
                ],
            })
            .then((html) => res.send(html))
            .catch((err) => next(err));
    });

    return server;
}

function run(): void {
    const port = process.env['PORT'] || 4000;

    // Start up the Node server
    const server = app();
    server.listen(port, () => {
        console.log(`Node Express server listening on http://localhost:${port}`);
    });
}

// Webpack will replace 'require' with '__webpack_require__'
// '__non_webpack_require__' is a proxy to Node 'require'
// The below code is to ensure that the server is run only when not requiring the bundle.
declare const __non_webpack_require__: NodeRequire;
const mainModule = __non_webpack_require__.main;
const moduleFilename = (mainModule && mainModule.filename) || '';
if (moduleFilename === __filename || moduleFilename.includes('iisnode')) {
    run();
}

export default AppServerModule;
```

blakeoxx commented:
Thanks for your kind words, @betojsx ! Just glad to help :)

> do you happen to have any working demos for Angular 17 with a workflow similar to mine?

This PR is intended for use with NG18 projects. I'm assuming there are some compatibility issues between the Angular internals. Have you tried using the v17 artifact referenced in [this comment](https://github.com/stevermeister/ngx-cookie-service/issues/266#issuecomment-2326675401)? If that doesn't resolve your issue, let me know and I'll be happy to create a demo repo for v17 as well.

betojsx commented:
> Have you tried using the v17 artifact referenced in https://github.com/stevermeister/ngx-cookie-service/issues/266#issuecomment-2326675401

Yep! I forgot to mention that I already have the v17 artifact in my project.

```json
"ngx-cookie-service-ssr": "github:blakeoxx/ngx-cookie-service-ssr-artifact#v17"
```

I really appreciate your effort to help. At this point, I’ve tried everything I could but still haven’t figured out how to make it work, so I’ll wait for some examples

blakeoxx commented:
Looking over your `server.ts` again, I originally missed that you don't have the response token in your providers. This token is necessary for the package to write cookies on the HTTP response sent to the client. Try updating your `server.ts` with the following:

```
import { REQUEST as SSR_REQUEST, RESPONSE as SSR_RESPONSE } from 'ngx-cookie-service-ssr';
// ...
commonEngine
            .render({
                // ...
                providers: [
                    { provide: APP_BASE_HREF, useValue: baseUrl },
                    { provide: SSR_REQUEST, useValue: req },
                    { provide: SSR_RESPONSE, useValue: res },
                    { provide: 'RESPONSE', useValue: res },
                    
                    // harcoded host for testing
                    {
                        provide: 'X_FORWARDED_HOST',
                        useValue: 'sub.mydomain.com',
                    },
                ],
            })
```

betojsx commented:
OMG @blakeoxx, you solved it!

That was exactly the issue! I can’t thank you enough. 🙏

I spent countless hours debugging this, even put the project on hold for a while out of frustration. I’m so glad I decided to share it here, and even more thankful for your help. You’ve saved me a ton of time!

blakeoxx commented:
@pavankjadda any updates on your review of this PR? Anything needed to keep this moving?

pavankjadda commented:
Haven't got time. Will try to do this week.

maxisam commented on projects/ngx-cookie-service-ssr/src/lib/ssr-cookie.service.ts:
```diff
@@ -54,6 +56,71 @@ export class SsrCookieService {
+      responseCookies = [responseCookies];
+    }
+
+    let allCookies = new Map(requestCookies);
```
is it necessary to create another map?

blakeoxx replied:
I think this was a leftover of an earlier implementation where I compared the `requestCookies` to the `allCookies` maps. Anyways, you're right, not needed anymore.

maxisam commented on projects/ngx-cookie-service-ssr/src/lib/ssr-cookie.service.ts:
```diff
@@ -54,6 +56,71 @@ export class SsrCookieService {
+      let [cookieName, cookieValue] = currentCookie.split('=');
+
+      // Remove any extra spaces from the beginning of cookie names. These are a side effect of browser/express cookie concatenation
+      cookieName = cookieName.replace(/^ +/, '');
```
IMO, trimStart might be more straightforward.

blakeoxx replied:
I used this to keep consistent with previous code, but agreed, happy to change to `trimStart`.

See https://github.com/stevermeister/ngx-cookie-service/pull/320/files#diff-4bd5ade27ff959a3d837687c0ae49f2f05cdabb3b655a61cd4c0a03497171c10L109

maxisam commented on projects/ngx-cookie-service-ssr/src/lib/ssr-cookie.service.ts:
```diff
@@ -64,9 +131,16 @@ export class SsrCookieService {
+      const regExp: RegExp = SsrCookieService.getCookieRegExp(name);
+      return regExp.test(this.document.cookie);
+    } else {
+      // Server-side cookie check considering incoming cookies from the request and already set cookies on the response
```
seems like we just need this part for check.

maxisam commented:
i wonder if we add code like below and access this.rawCookie instead of this.document.cookie

```typescript
  get rawCookie(): string | undefined {
    if (this.documentIsAccessible) {
      return this.document.cookie;
    } else {
      const responseCookies = this.response?.get('Set-Cookie');
      let respCookies = '';
      if (responseCookies) {
        respCookies = Array.isArray(responseCookies) ? responseCookies.join(';') : responseCookies;
        return this.request?.headers.cookie + ';' + respCookies;
      } else {
        return this.request?.headers.cookie;
      }
    }
  }
  ```
  We probably only need change Set method.

blakeoxx commented:
> i wonder if we add code like below and access this.rawCookie instead of this.document.cookie
> 
> ```ts
>   get rawCookie(): string | undefined {
>     if (this.documentIsAccessible) {
>       return this.document.cookie;
>     } else {
>       const responseCookies = this.response?.get('Set-Cookie');
>       let respCookies = '';
>       if (responseCookies) {
>         respCookies = Array.isArray(responseCookies) ? responseCookies.join(';') : responseCookies;
>         return this.request?.headers.cookie + ';' + respCookies;
>       } else {
>         return this.request?.headers.cookie;
>       }
>     }
>   }
> ```
> 
> We probably only need change Set method.

@maxisam I think the only issue I see with this code is the case of the cookie being set in the request as well as response. In this code, the cookie would be duplicated, causing the first instance from the request to be the value always returned by `this.get`. This is the main reason why I elected to use a key-value Map in `this.getCombinedCookies`. I also think names and values of cookies from the request/response headers may be URI encoded since they may be returned by the express interface as one concatenated string.

blakeoxx commented:
I do think switching the order would fix the issue. However, I'd caution that it might be surprising behavior to future maintainers that cookies appear multiple times in the string and only the first occurrence is used due to the specific coding of `this.get`.

In the end, I also don't feel that strongly about either implementation. If you feel it would ease maintainability, I'll happily update. Do you think the maintainability boost your solution offers is worth the surprising behavior?

maxisam commented:
well, this will be @pavankjadda 's decide. Either way is fine with me, but i do think having one solution is probably better than having both.

danpoland commented:
With the release [RESPONSE_INIT](https://angular.dev/api/core/RESPONSE_INIT) there isn't a need for express specific objects.

 I threw together a quick example to test out using RESPONSE_INIT: https://github.com/danpoland/ngx-cookie-service/commit/9cf72876008f65cb1bc21ba6abad5c6859cde391

REPONSE_INIT is still in developer preview but my change works fine.
